flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcUtils.java

```diff
@@ -0,0 +1,2657 @@
+				}
+				else { // set list of list as an field on Row
+					for (int i = 0; i < rows.length; i++) {
+						length = (int) lengthVector[i];
+						if (length == 0) {
+							((Row) rows[i]).setField(fieldIdx, new Object[0]);
+						}
+						else {
+							temp = new Object[length];
+							System.arraycopy(nestedList, offset, temp, 0, length);
+							offset = offset + length;
+							((Row) rows[i]).setField(fieldIdx, temp);
+						}
+					}
+				}
+			}
+		}
+
+	}
+
+	private static void readNonNullMapColumn(Object[] rows, int fieldIdx, MapColumnVector map, TypeDescription schema, long[] lengthVector) {
+
+		List<TypeDescription> fieldType = schema.getChildren();
+		TypeDescription keyType = fieldType.get(0);
+		TypeDescription valueType = fieldType.get(1);
+
+		if (map.isRepeating) {
+			// TODO: <Need Feedback> but I think don't need in case of struct, list and map.
+			// need to remove this condition
+		} else {
+
+			ColumnVector keys = map.keys;
+			ColumnVector values = map.values;
+			Object[] keyRows = new Object[rows.length];
+			Object[] valueRows = new Object[rows.length];
+			for (int i = 0; i < rows.length; i++) {
+				keyRows[i] = new Object();
+				valueRows[i] = new Object();
+			}
+
+			// read map kes and values
+			readField(keyRows, -1, keyType, keys, null);
+			readField(valueRows, -1, valueType, values, null);
+
+			if (lengthVector == null) {
+				long[] lengthVectorMap = map.lengths;
+				int offset = 0;
+				if (fieldIdx == -1) {
+					for (int i = 0; i < rows.length; i++) {
+						HashMap<Object, Object> resultMap = new HashMap<Object, Object>();
+						for (int j = 0; j < lengthVectorMap[i]; j++) {
+							resultMap.put(keyRows[offset], valueRows[offset]);
+							offset++;
+						}
+						rows[i] = resultMap;
+					}
+				}
+				else {
+					for (int i = 0; i < rows.length; i++) {
+						HashMap<Object, Object> resultMap = new HashMap<Object, Object>();
+						for (int j = 0; j < lengthVectorMap[i]; j++) {
+							resultMap.put(keyRows[offset], valueRows[offset]);
+							offset++;
+						}
+						((Row) rows[i]).setField(fieldIdx, resultMap);
+					}
+				}
+			}
+			else { // list of map
+
+				Object[] nestedMap = new Object[rows.length];
+				for (int i = 0; i < rows.length; i++) {
+					nestedMap[i] = new Object();
+				}
+
+				long[] lengthVectorMap = map.lengths;
+				int offset = 0;
+				for (int i = 0; i < rows.length; i++) {
+					HashMap<Object, Object> resultMap = new HashMap<Object, Object>();
+					for (int j = 0; j < lengthVectorMap[i]; j++) {
+						resultMap.put(keyRows[offset], valueRows[offset]);
+						offset++;
+					}
+					nestedMap[i] = resultMap;
+				}
+
+				offset = 0;
+				int length;
+				Object[] temp;
+				if (fieldIdx == -1) {
+					for (int i = 0; i < rows.length; i++) {
+						length = (int) lengthVector[i];
+						if (length == 0) {
+							rows[i] = new Object[0];
+						} else {
+							temp = new Object[length];
+							System.arraycopy(nestedMap, offset, temp, 0, length);
```

Comment:
It would be better to directly create the maps here instead of using an intermediate array.